### PR TITLE
Add rolling computations for dask.dataframe

### LINKED ISSUE #487: Rolling operations on dask.dataframe
This probably requires a simpler version of `da.ghost`

http://pandas.pydata.org/pandas-docs/stable/computation.html#moving-rolling-statistics-moments

### PR DESCRIPTION
**Pull request by jcrist**
Fixes #487.

### CODE REVIEW

File: dask/dataframe/rolling.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    applied = func(combined, window, *args)
+    return applied[extra:]
+
+
+def wrap_rolling(func):
+    """Create a chunked version of a pandas.rolling_* function"""
+    @wraps(func)
+    def rolling(arg, window, *args, **kwargs):
+        old_name = arg._name
+        new_name = next(tokens)
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: Can we provide a friendly name like `rolling_sum-43` rather than `-43`

File: dask/dataframe/rolling.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        f = partial(func, **kwargs)
```
Comment (mrocklin) [inline dask/dataframe/rolling.py]: I wonder if `rolling_chunk` could take `args` rather than `*args` and also take a straight `kwargs` dict (rather than `**kwargs`).  Lately I've been finding this approach to be more debuggable.

File: dask/dataframe/rolling.py (hunk 1 above)
Comment (jcrist) [inline dask/dataframe/rolling.py]: I thought about doing that, but decided against it to keep things simple. Easy fix though, sure.

File: dask/dataframe/tests/test_rolling.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    eq(pd.rolling_std(p, 3), dd.rolling_std(d, 3))
+    eq(pd.rolling_var(p, 3), dd.rolling_var(d, 3))
+    eq(pd.rolling_skew(p, 3), dd.rolling_skew(d, 3))
+    eq(pd.rolling_kurt(p, 3), dd.rolling_kurt(d, 3))
+    eq(pd.rolling_quantile(p, 3, 0.5), dd.rolling_quantile(d, 3, 0.5))
+    mad = lambda x: np.fabs(x - x.mean()).mean()
+    eq(pd.rolling_apply(p, 3, mad), dd.rolling_apply(d, 3, mad))
+    eq(pd.rolling_window(p, 3, 'boxcar'), dd.rolling_window(d, 3, 'boxcar'))
+    # Test with kwargs
+    eq(pd.rolling_sum(p, 3, how='min'), dd.rolling_sum(d, 3, how='min'))
```
Comment (mrocklin) [inline dask/dataframe/tests/test_rolling.py]: `how='min'` seems odd for `rolling_sum`

**Comment by mrocklin** [comment]
@cpcloud can you give this a quick check for Pandas correctness?

File: dask/dataframe/rolling.py (hunk 2 above)
Comment (jcrist) [inline dask/dataframe/rolling.py]: That's what I had original, but that failed because `rolling_apply` takes a function, which resulted in an args tuple starting with a function. Could certainly make that a list though, idk.

File: dask/dataframe/tests/test_rolling.py (hunk 3 above)
Comment (jcrist) [inline dask/dataframe/tests/test_rolling.py]: I just picked a kwarg and a function. More realistic suggestions would be appreciated.

File: dask/dataframe/rolling.py (hunk 1 above)
Comment (mrocklin) [inline dask/dataframe/rolling.py]: Ah, I see now.  I've been taking `tokens` from core these days and have eliminated most of the `read_csv_names, ...` iterators.  Not sure if this is best or not.

File: dask/dataframe/rolling.py (hunk 2 above)
Comment (mrocklin) [inline dask/dataframe/rolling.py]: Oh funny.  What a tricky corner case.

File: dask/dataframe/rolling.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import pandas as pd
+
+tokens = ('rolling-%d' % i for i in count(1))
+
+
+def rolling_chunk(func, part1, part2, window, *args):
+    if part1.shape[0] < window:
+        raise NotImplementedError("Window larger than partition size")
+    extra = window - 1
+    combined = pd.concat([part1[-extra:], part2])
```
Comment (cpcloud) [inline dask/dataframe/rolling.py]: you'll want to use `iloc` here and also on `applied`

File: dask/dataframe/rolling.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+from toolz import merge
+import pandas as pd
+
+tokens = ('rolling-%d' % i for i in count(1))
+
+
+def rolling_chunk(func, part1, part2, window, *args):
+    if part1.shape[0] < window:
+        raise NotImplementedError("Window larger than partition size")
+    extra = window - 1
```
Comment (cpcloud) [inline dask/dataframe/rolling.py]: i think this should be `window + 1` and not `window - 1`

if `extra == window - 1` then a window size of 1 will concat each chunk to the next one and then doing a rolling window on it, with `window == 0` the results should be all `nan` the size of the `series` and currently we get all `nan` truncated to the size of `npartitions + len(chunk)`. While these *are* edges case, pandas handles it correctly.

Comment (cpcloud) [inline dask/dataframe/rolling.py]: also needs a test against window sizes 0 and 1

File: dask/dataframe/rolling.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+def wrap_rolling(func):
+    """Create a chunked version of a pandas.rolling_* function"""
+    @wraps(func)
+    def rolling(arg, window, *args, **kwargs):
+        if not isinstance(window, int):
+            raise TypeError('Window must be an integer')
+        if window < 0:
+            raise ValueError('Window must be a positive integer')
+        if 'freq' in kwargs:
+            raise NotImplementedError('"freq" kwarg for rolling functions')
```
Comment (cpcloud) [inline dask/dataframe/rolling.py]: i'd probably make this

```python
if 'freq' in kwargs or 'how' in kwargs:
    raise NotImplementedError("resampling before rolling computations not supported")
```
to remind us what those args do

Comment (cpcloud) [inline dask/dataframe/rolling.py]: or just explicitly name them and make them both None and if they deviate then raise the above error

File: dask/dataframe/rolling.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def wrap_rolling(func):
+    """Create a chunked version of a pandas.rolling_* function"""
+    @wraps(func)
+    def rolling(arg, window, *args, **kwargs):
+        if not isinstance(window, int):
+            raise TypeError('Window must be an integer')
+        if window < 0:
+            raise ValueError('Window must be a positive integer')
```
Comment (cpcloud) [inline dask/dataframe/rolling.py]: pandas will actually catch this, but if dask is going to do it we should do it during dask construction, not during the computation

File: dask/dataframe/rolling.py (hunk 6 above)
Comment (jcrist) [inline dask/dataframe/rolling.py]: woops, missed `how` as well.

File: dask/dataframe/rolling.py (hunk 7 above)
Comment (cpcloud) [inline dask/dataframe/rolling.py]: therefore saving users a lot of time

File: dask/dataframe/tests/test_rolling.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    eq(pd.rolling_kurt(p, 3), dd.rolling_kurt(d, 3))
+    eq(pd.rolling_quantile(p, 3, 0.5), dd.rolling_quantile(d, 3, 0.5))
+    mad = lambda x: np.fabs(x - x.mean()).mean()
+    eq(pd.rolling_apply(p, 3, mad), dd.rolling_apply(d, 3, mad))
+    eq(pd.rolling_window(p, 3, 'boxcar'), dd.rolling_window(d, 3, 'boxcar'))
+    # Test with edge-case window sizes
+    eq(pd.rolling_sum(p, 0), dd.rolling_sum(d, 0))
+    eq(pd.rolling_sum(p, 1), dd.rolling_sum(d, 1))
+    # Test with kwargs
+    eq(pd.rolling_sum(p, 3, how='min'), dd.rolling_sum(d, 3, how='min'))
```
Comment (cpcloud) [inline dask/dataframe/tests/test_rolling.py]: maybe these should be in a separate test that tests that they fail

File: dask/dataframe/rolling.py (hunk 7 above)
Comment (jcrist) [inline dask/dataframe/rolling.py]: This is during dask construction.

Comment (cpcloud) [inline dask/dataframe/rolling.py]: oh ... duh my mistake

**Comment by jcrist** [comment]
Ok, I *think* this is good to go now.

**Comment by cpcloud** [comment]
LGTM
